napari/_qt/_tests/test_qt_viewer.py

```diff
@@ -675,3 +679,123 @@ def test_create_non_empty_viewer_model(qtbot):
+
+
+@pytest.mark.sync_only
+def test_async_slice_image_on_ndisplay_change(make_napari_viewer, qtbot):
+    np.random.seed(0)
+    data = np.random.rand(3, 4, 5)
+    viewer = make_napari_viewer()
+    vispy_image = setup_viewer_for_async_slice_image(viewer, data)
+    assert viewer.dims.ndisplay != 3
+
+    viewer.dims.ndisplay = 3
+
+    wait_until_vispy_image_data_equal(qtbot, vispy_image, data)
+
+
+@pytest.mark.sync_only
+def test_async_slice_multiscale_image_on_pan(make_napari_viewer, qtbot):
+    viewer = make_napari_viewer()
+    np.random.seed(0)
+    data = [np.random.rand(4, 8, 10), np.random.rand(2, 4, 5)]
+    vispy_image = setup_viewer_for_async_slice_image(viewer, data)
+
+    # Check that we're initially slicing the middle of the first dimension
+    # over the whole of lowest resolution image.
+    assert viewer.dims.not_displayed == (0,)
+    assert viewer.dims.current_step[0] == 2
+    image = vispy_image.layer
+    assert image._data_level == 1
+    np.testing.assert_equal(image.corner_pixels, [[0, 0, 0], [0, 4, 5]])
+
+    # Simulate panning to the left by changing the corner pixels in the last
+    # dimension, which corresponds to x/columns, then triggering a reslice.
+    image.corner_pixels = np.array([[0, 0, 0], [0, 4, 3]])
+    image.events.reslice(Event('reslice', layer=image))
+
+    wait_until_vispy_image_data_equal(qtbot, vispy_image, data[1][1, 0:4, 0:3])
+
+
+@pytest.mark.sync_only
+def test_async_slice_multiscale_image_on_zoom(qtbot, make_napari_viewer):
+    viewer = make_napari_viewer()
+    np.random.seed(0)
+    data = [np.random.rand(4, 8, 10), np.random.rand(2, 4, 5)]
+    vispy_image = setup_viewer_for_async_slice_image(viewer, data)
+
+    # Check that we're initially slicing the middle of the first dimension
+    # over the whole of lowest resolution image.
+    assert viewer.dims.not_displayed == (0,)
+    assert viewer.dims.current_step[0] == 2
+    image = vispy_image.layer
+    assert image._data_level == 1
+    np.testing.assert_equal(image.corner_pixels, [[0, 0, 0], [0, 4, 5]])
+
+    # Simulate zooming into the middle of the higher resolution image.
+    image._data_level = 0
+    image.corner_pixels = np.array([[0, 2, 3], [0, 6, 7]])
+    image.events.reslice(Event('reslice', layer=image))
+
+    wait_until_vispy_image_data_equal(qtbot, vispy_image, data[0][2, 2:6, 3:7])
+
+
+def setup_viewer_for_async_slice_image(
+    viewer: Viewer,
+    data: Union[np.ndarray, List[np.ndarray]],
+) -> VispyImageLayer:
+    # Initially force synchronous slicing so any slicing caused
+    # by adding the image finishes before any other slicing starts.
+    viewer._layer_slicer._force_sync = True
+    # Add the image and get the corresponding vispy image.
+    layer = viewer.add_image(data)
+    vispy_layer = viewer.window._qt_viewer.layer_to_visual[layer]
+    # Then allow asynchronous slicing for testing.
+    viewer._layer_slicer._force_sync = False
```

Comment:
I think that's a lot of tests, so I'd personally prefer to wait until we actually flip this flag before handling that. If there are strong feelings about this one, I'll likely make all the changes at the end to avoid cluttering the file changes during review.

Reply:
Agreed.